python/cudf_polars/cudf_polars/experimental/parallel.py

```diff
@@ -234,3 +255,61 @@ def evaluate_dask(ir: IR) -> DataFrame:
 
     graph, key = task_graph(ir, partition_info)
     return get(graph, key)
+
+
+def _concat(dfs: Sequence[DataFrame]) -> DataFrame:
+    # Concatenate a sequence of DataFrames vertically
+    return Union.do_evaluate(None, *dfs)
+
+
+##
+## Union
+##
+
+
+@lower_ir_node.register(Union)
+def _(
+    ir: Union, rec: LowerIRTransformer
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    # zlice must be None
+    if ir.zlice is not None:
+        return rec.state["default_mapper"](ir)  # pragma: no cover
+
+    # Lower children
+    children, _partition_info = zip(*(rec(c) for c in ir.children), strict=False)
+    partition_info = reduce(operator.or_, _partition_info)
+
+    # Partition count is the sum of all child partitions
+    count = sum(partition_info[c].count for c in children)
+
+    # Return reconstructed node and partition-info dict
+    new_node = ir.reconstruct(children)
+    partition_info[new_node] = PartitionInfo(count=count)
+    return new_node, partition_info
+
+
+@generate_ir_tasks.register(Union)
+def _(
+    ir: Union, partition_info: MutableMapping[IR, PartitionInfo]
+) -> MutableMapping[Any, Any]:
+    part_out = 0
+    key_name = get_key_name(ir)
+    graph: MutableMapping[Any, Any] = {}
+    for child in ir.children:
+        for i in range(partition_info[child].count):
+            graph[(key_name, part_out)] = (get_key_name(child), i)
+            part_out += 1
+    return graph
```

Comment:
```suggestion
    partition = itertools.count()
    graph = {
        (key_name, next(partition)): (get_key_name(child), i)
        for i in range(partition_info[child].count)
        for child in ir.children
    }
    return graph
```

This iteration over partition counts is quite common, I wonder if we should make:

```
class PartitionCount:
    def __init__(self, key_name: str, count: int):
        self.key_name = key_name
        self.count = count
    def __iter__(self):
        yield from ((self.key_name, i) for i in range(self.count))
```

Then we could write:

```
partition = itertools.count()
graph = {
    (key_name, next(partition)): dep
    for dep in partition_info[child]
    for child in ir.children
}
```

Reply:
I'm not sure I completely understand the suggestion. In your `class PartitionCount` sketch, do you mean `class PartitionInfo`?

Reply:
I'd like to keep the IR object separate from the PartitionInfo object. How about:

```
class PartitionInfo:
   ...
    def keys(self, node: Node) -> Iterator[tuple[str, int]]:
        """Return the partitioned keys for a given node."""
        name = get_key_name(node)
        yield from ((name, i) for i in range(self.count))
```

Reply:
Latest version of the branch adds/uses the `keys` method for convenience.